Importing all needed Python modules

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import datetime as dt
import seaborn as sns
sns.set(style="darkgrid")

In [3]:
import datetime
from datetime import datetime

In [4]:
# Used to execute multple code lines in a single cell
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

Importing Raw Data Files

In [5]:
ffx = pd.read_csv("ffxLighting_original_june28.txt")

In [6]:
ffx2 = pd.read_csv("ffxLighting_new_july1.txt")

In [7]:
ffx3 = pd.read_csv("gray_file_july1.txt")

Initial Look At The Data

In [8]:
ffx.head()
ffx2.head()
ffx3.head()

count                      timings
0      1   2020-06-28 16:40:11.538763
1      2   2020-06-28 16:40:17.543537
2      3   2020-06-28 16:40:28.666688
3      4   2020-06-28 16:40:34.640885
4      5   2020-06-28 16:40:45.889237

count                      timings
0      1   2020-07-01 22:18:22.806651
1      2   2020-07-01 22:18:49.157322
2      3   2020-07-01 22:18:55.197844
3      4   2020-07-01 22:19:06.378970
4      5   2020-07-01 22:19:12.418819

grayscale                      timings
0  235.556693   2020-07-01 22:18:21.991453
1  235.544079   2020-07-01 22:18:22.187393
2  229.834134   2020-07-01 22:18:22.202118
3  229.882272   2020-07-01 22:18:22.219390
4  229.889206   2020-07-01 22:18:22.234170

In [9]:
ffx["count"][0]
ffx[" timings"][0]

1

' 2020-06-28 16:40:11.538763'

In [10]:
type(ffx["count"][0])
type(ffx[" timings"][0])

numpy.int64

str

In [11]:
ffx2["count"][0]
ffx2[" timings"][0]

1

' 2020-07-01 22:18:22.806651'

In [12]:
ffx3["grayscale"][0]
ffx3[" timings"][0]

235.55669270833332

' 2020-07-01 22:18:21.991453'

Getting General Info About Raw Data (NOTE: The Stupid Space!!!)

In [13]:
ffx.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1135 entries, 0 to 1134
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   count     1135 non-null   int64 
 1    timings  1135 non-null   object
dtypes: int64(1), object(1)
memory usage: 17.9+ KB


In [14]:
ffx2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1859 entries, 0 to 1858
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   count     1859 non-null   int64 
 1    timings  1859 non-null   object
dtypes: int64(1), object(1)
memory usage: 29.2+ KB


In [15]:
ffx3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 399539 entries, 0 to 399538
Data columns (total 2 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   grayscale  399539 non-null  float64
 1    timings   399539 non-null  object 
dtypes: float64(1), object(1)
memory usage: 6.1+ MB


In [16]:
# All date/time column data is "str" !!!
type(ffx[" timings"][0])
type(ffx2[" timings"][0])
type(ffx3[" timings"][0])

str

str

str

# **Cleaning Data for ffx (original data)**

Need to first get rid of the stupid space (in column header)!

In [17]:
ffx.head()
ffx.columns
ffx[" timings"][0]

count                      timings
0      1   2020-06-28 16:40:11.538763
1      2   2020-06-28 16:40:17.543537
2      3   2020-06-28 16:40:28.666688
3      4   2020-06-28 16:40:34.640885
4      5   2020-06-28 16:40:45.889237

Index(['count', ' timings'], dtype='object')

' 2020-06-28 16:40:11.538763'

In [18]:
ffx.rename(mapper={" timings": "timings"}, axis=1, inplace=True)
ffx.columns

Index(['count', 'timings'], dtype='object')

Need to get rid of spaces in the "timings" DATA as well!!!

In [19]:
ffx["timings"][0]
type(ffx["timings"][0])

' 2020-06-28 16:40:11.538763'

str

In [20]:
ffx["timings"] = ffx["timings"].apply(lambda string: string[1:])

In [21]:
ffx["timings"][0]
type(ffx["timings"][0])

'2020-06-28 16:40:11.538763'

str

In [22]:
ffx.head()

count                     timings
0      1  2020-06-28 16:40:11.538763
1      2  2020-06-28 16:40:17.543537
2      3  2020-06-28 16:40:28.666688
3      4  2020-06-28 16:40:34.640885
4      5  2020-06-28 16:40:45.889237

Look for null values? (I already know there aren't any)

In [23]:
ffx.count()
ffx.isnull().count()

count      1135
timings    1135
dtype: int64

count      1135
timings    1135
dtype: int64

I happen to know that some values at the tail of the dataframe is not needed (it was recorded while I was dismantling the data capture setup). We will find the bad data, and remove the respective rows.

Based on the project, we will most likely be more interested in time differences between lightning strikes.

**SO! We must first change the timings data to 'datetime/timedelta' objects**

In [24]:
ffx.head()
ffx["timings"][0]
type(ffx["timings"][0])

count                     timings
0      1  2020-06-28 16:40:11.538763
1      2  2020-06-28 16:40:17.543537
2      3  2020-06-28 16:40:28.666688
3      4  2020-06-28 16:40:34.640885
4      5  2020-06-28 16:40:45.889237

'2020-06-28 16:40:11.538763'

str

In [25]:
ffx["Datetime"] = pd.to_datetime(ffx["timings"])
ffx.head()
ffx["Datetime"][0]
type(ffx["Datetime"][0])

count                     timings                   Datetime
0      1  2020-06-28 16:40:11.538763 2020-06-28 16:40:11.538763
1      2  2020-06-28 16:40:17.543537 2020-06-28 16:40:17.543537
2      3  2020-06-28 16:40:28.666688 2020-06-28 16:40:28.666688
3      4  2020-06-28 16:40:34.640885 2020-06-28 16:40:34.640885
4      5  2020-06-28 16:40:45.889237 2020-06-28 16:40:45.889237

Timestamp('2020-06-28 16:40:11.538763')

pandas._libs.tslibs.timestamps.Timestamp

Now that everything is converted to actual "datetime" dtype, we can look for our bad data. It's somewhere at the end, remember!

In [26]:
ffx[1100:]

count                     timings                   Datetime
1100   1101  2020-06-28 19:17:45.300614 2020-06-28 19:17:45.300614
1101   1102  2020-06-28 19:17:51.295674 2020-06-28 19:17:51.295674
1102   1103  2020-06-28 19:18:02.442681 2020-06-28 19:18:02.442681
1103   1104  2020-06-28 19:18:08.487168 2020-06-28 19:18:08.487168
1104   1105  2020-06-28 19:18:19.587722 2020-06-28 19:18:19.587722
1105   1106  2020-06-28 19:18:25.632636 2020-06-28 19:18:25.632636
1106   1107  2020-06-28 19:18:36.770161 2020-06-28 19:18:36.770161
1107   1108  2020-06-28 19:18:42.812585 2020-06-28 19:18:42.812585
1108   1109  2020-06-28 19:18:53.944019 2020-06-28 19:18:53.944019
1109   1110  2020-06-28 19:18:59.994721 2020-06-28 19:18:59.994721
1110   1111  2020-06-28 19:19:11.182394 2020-06-28 19:19:11.182394
1111   1112  2020-06-28 19:19:17.140761 2020-06-28 19:19:17.140761
1112   1113  2020-06-28 19:19:28.370228 2020-06-28 19:19:28.370228
1113   1114  2020-06-28 19:19:34.304837 2020-06-28 19:19:34.304837
1114   1115  2020-06-28 19:19:45.509329 2020-06-28 19:19:45.509329
1115   1116  2020-06-28 19:19:51.529480 2020-06-28 19:19:51.529480
1116   1117  2020-06-28 19:20:02.690825 2020-06-28 19:20:02.690825
1117   1118  2020-06-28 19:20:08.719560 2020-06-28 19:20:08.719560
1118   1119  2020-06-28 19:20:18.647719 2020-06-28 19:20:18.647719
1119   1120  2020-06-28 19:20:23.351210 2020-06-28 19:20:23.351210
1120   1121  2020-06-28 19:20:40.833734 2020-06-28 19:20:40.833734
1121   1122  2020-06-28 19:20:46.595614 2020-06-28 19:20:46.595614
1122   1123  2020-06-28 19:20:48.694830 2020-06-28 19:20:48.694830
1123   1124  2020-06-28 19:21:02.183183 2020-06-28 19:21:02.183183
1124   1125  2020-06-28 19:21:09.733099 2020-06-28 19:21:09.733099
1125   1126  2020-06-28 19:21:17.263125 2020-06-28 19:21:17.263125
1126   1127  2020-06-28 19:21:30.427996 2020-06-28 19:21:30.427996
1127   1128  2020-06-28 19:21:45.578473 2020-06-28 19:21:45.578473
1128   1129  2020-06-28 19:21:45.679359 2020-06-28 19:21:45.679359
1129   1130  2020-06-28 19:21:56.824975 2020-06-28 19:21:56.824975
1130   1131  2020-06-28 19:23:27.663365 2020-06-28 19:23:27.663365
1131   1132  2020-06-28 19:24:52.651587 2020-06-28 19:24:52.651587
1132   1133  2020-06-28 19:24:58.391024 2020-06-28 19:24:58.391024
1133   1134  2020-06-28 19:25:35.689479 2020-06-28 19:25:35.689479
1134   1135  2020-06-28 19:25:41.456840 2020-06-28 19:25:41.456840

In [49]:
# count_column = f'{ffx["count"][i+1]}---{ffx["count"][i]}'
# time_diff_column = ffx["Datetime"][i+1]-ffx["Datetime"][i]

In [50]:
bad_time_diff = pd.DataFrame([(f'{ffx["count"][i+1]}---{ffx["count"][i]}', ffx["Datetime"][i+1]-ffx["Datetime"][i]) for i in range(1100,1134)])

In [51]:
bad_time_diff

0               1
0   1102---1101 00:00:05.995060
1   1103---1102 00:00:11.147007
2   1104---1103 00:00:06.044487
3   1105---1104 00:00:11.100554
4   1106---1105 00:00:06.044914
5   1107---1106 00:00:11.137525
6   1108---1107 00:00:06.042424
7   1109---1108 00:00:11.131434
8   1110---1109 00:00:06.050702
9   1111---1110 00:00:11.187673
10  1112---1111 00:00:05.958367
11  1113---1112 00:00:11.229467
12  1114---1113 00:00:05.934609
13  1115---1114 00:00:11.204492
14  1116---1115 00:00:06.020151
15  1117---1116 00:00:11.161345
16  1118---1117 00:00:06.028735
17  1119---1118 00:00:09.928159
18  1120---1119 00:00:04.703491
19  1121---1120 00:00:17.482524
20  1122---1121 00:00:05.761880
21  1123---1122 00:00:02.099216
22  1124---1123 00:00:13.488353
23  1125---1124 00:00:07.549916
24  1126---1125 00:00:07.530026
25  1127---1126 00:00:13.164871
26  1128---1127 00:00:15.150477
27  1129---1128 00:00:00.100886
28  1130---1129 00:00:11.145616
29  1131---1130 00:01:30.838390
30  1132---1131 00:01:24.988222
31  1133---1132 00:00:05.739437
32  1134---1133 00:00:37.298455
33  1135---1134 00:00:05.767361

We can see in the above datafram that the time differences stop following the pattern starting with index 17. This means that the datapoints with count of **1119 and onward** are the bad data.

So, in our ffx dataframe, let us now delete the rows with 'count' starting at 1119 and greater. *This translates to indices of **1118** and greater!*

In [56]:
ffx.shape

(1135, 3)

In [59]:
ffx_cleaned = ffx.drop(ffx[ffx["count"] >= 1119].index, axis=0)

In [61]:
ffx_cleaned

count                     timings                   Datetime
0         1  2020-06-28 16:40:11.538763 2020-06-28 16:40:11.538763
1         2  2020-06-28 16:40:17.543537 2020-06-28 16:40:17.543537
2         3  2020-06-28 16:40:28.666688 2020-06-28 16:40:28.666688
3         4  2020-06-28 16:40:34.640885 2020-06-28 16:40:34.640885
4         5  2020-06-28 16:40:45.889237 2020-06-28 16:40:45.889237
...     ...                         ...                        ...
1113   1114  2020-06-28 19:19:34.304837 2020-06-28 19:19:34.304837
1114   1115  2020-06-28 19:19:45.509329 2020-06-28 19:19:45.509329
1115   1116  2020-06-28 19:19:51.529480 2020-06-28 19:19:51.529480
1116   1117  2020-06-28 19:20:02.690825 2020-06-28 19:20:02.690825
1117   1118  2020-06-28 19:20:08.719560 2020-06-28 19:20:08.719560

[1118 rows x 3 columns]

# **Cleaning Data for ffx2 (newer data)**

In [ ]:
ffx.shape

In [ ]:
counts = ffx['count,']
counts

In [ ]:
type(counts[1])

In [ ]:
timings = ffx['timings']
timings.head()

In [ ]:
# type(timings) #--> Series
type(timings[1]) #--> String

In [ ]:
timings2 = pd.to_datetime(timings)
timings2.head()

In [ ]:
# type(timings2[1]) #--> Timestamp
# type(timings2[1]-timings2[0]) #--> Timedelta
timings2[10]-timings2[0]

In [ ]:
(timings2[10]-timings2[0]).seconds + (timings2[10]-timings2[0]).microseconds/10**6

In [ ]:
time_diff = [(timings2[i+1]-timings2[i]).seconds + (timings2[i+1]-timings2[i]).microseconds/10**6 for i in range(0, 1134)]
# len(time_diff) #--> 1134
time_diff[1115:]

In [ ]:
# Need to remove last 17 entries of time_diff list
y = time_diff[:len(time_diff)-17]
y[1110:]
# len(y) #-->1117

In [ ]:
# Removing last 17 corresponding entries in counts list
# and an extra one to make it the same size as time_diff

x = counts[:len(counts)-18]
len(x)

In [ ]:
# fig = plt.figure()

# axes = fig.add_axes([0,0,1,1])
# axes.set_xlabel("Counts")
# axes.set_ylabel("Time Differences")

# # axes.plot(x, y)


In [ ]:
plt.scatter(x,y)

In [ ]:
average_time = sum(y)/len(y)
average_time

In [ ]:
above_average = [time for time in y if time > average_time]
above_average[:5]
# len(above_average) #--> 558

In [ ]:
below_average = [time for time in y if time < average_time]
below_average[:5]
# len(below_average) #--> 559

In [ ]:
count_above = [count for (count, time) in zip(x,y) if time > average_time]
count_above # --> ALL EVEN NUMBERS!!
len(count_above)

In [ ]:
avg_of_above = sum(above_average)/len(above_average)
avg_of_above

In [ ]:
std_above = np.std(above_average)
std_above

In [ ]:
avg_of_below = sum(below_average)/len(below_average)
avg_of_below

In [ ]:
std_below = np.std(below_average)
std_below

In [ ]:
count_below = [count for (count, time) in zip(x,y) if time < average_time]
count_below # --> ALL EVEN NUMBERS!!
len(count_below)

In [ ]:
plt.scatter(count_above, above_average)

In [ ]:
plt.scatter(count_below, below_average)

In [ ]:
num_bins = 20
plt.hist(above_average, bins = num_bins)

In [ ]:
num_bins = 20
plt.hist(below_average, bins = num_bins)

In [ ]:
count_above[:5]

In [ ]:
def all_even_or_odd(array):

    remainders = list(map(lambda x: int(x[:-1]) % 2, array))
    
    if remainders.count(1) == len(array):
        return "The array contains all odd integers"
    elif remainders.count(0) == len(array):
        return "The array contains all even integers"
    else:
        return "The array contains both even and odd integers"

In [ ]:
all_even_or_odd(count_above)

In [ ]:
all_even_or_odd(count_below)